### ISSUE TITLE
Extending native `string`

### ISSUE THREAD

**Issue by andyshuxin** [Microsoft/TypeScript#9212]
In short, is it possible to make a `string` with its own method?

Something like
```
color.toString() === `hsl(120, 20%, 50%)` // true
color.h // 120
color.s // 0.2
color.l  // 0.5
color.lighten(0.5).toJString() === `hsl(120, 20%, 100%)` // true

function print(content: string) {
  console.info(content)
}

print(s) // no semantic error here
```

I think, since `'str'.toString()` works, it might make sense to attach even more methods to the string?

It doesn't actually need to be a `string`, as long as it has a static type compatible with `string` and after converted to `string` it has the intended value.

**Comment by blendsdk** [comment]
Is this what you are looking for? [Check towards the end of the file](https://github.com/blendsdk/material-blend/blob/master/blend/src/common/Utils.ts)

**Comment by andyshuxin** [comment]
@blendsdk Not exactly. I'm not trying to attach methods to all strings. I try to make an object that has a static type of `string`. It'll ideally affect only a few objects, not every instance of string.

**Comment by kitsonk** [comment]
You would do it like you would in JavaScript, by augmenting the `String.prototype`.

If you want to modify the interface `String`, interfaces are open ended:

```typescript
interface String {
    h: number;
}
```

Of course augmenting globals like this is generally considered a bad design pattern though and subclassing built-ins would be better.

**Comment by aleksey-bykov** [comment]
you can extend a global interface locally

```typescript
declare global {
    interface String {
        myMegaMethod(): void;
    }
}    
'hey'.myMegaMethod(); // works
```
thank to https://github.com/Microsoft/TypeScript/issues/4166 that was shipped with TypeScript 1.8
